tests/basic/labels.rs

```diff
@@ -0,0 +1,337 @@
+// tasks.  We test two scenarios:
+// (1) the parent sets up a `LabelFn` to assign the child names at creation time
+//     (so that the names are in effect as soon as the child is created)
+// (2) the child tasks assign names to themselves by calling `set_task_name` as the
+//     first statement in their code,
+//
+// We use a custom tracing subscriber to monitor the list of `enabled` tasks that
+// is logged by Shuttle at each scheduling point.  (This also allows us to check
+// that Shuttle uses the user-assigned `TaskName` when generating debug output.)
+//
+// The two tests below check that:
+// - In scenario (1), the only names seen in the `runnable` list are `main-thread(0)`
+//   or names starting with `Child(`
+// - In scenario (2) the `runnable` list contains other names
+
+async fn label_fn_inner(set_name_before_spawn: bool) {
+    // Spawn 3 children
+    let handles = (0..3).map(|_| {
+        if set_name_before_spawn {
+            // Test scenario (2)
+            set_label_for_task(
+                me(),
+                LabelFn(Arc::new(|labels| {
+                    labels.insert(TaskName::from("Child"));
+                })),
+            );
+        }
+        future::spawn(async move {
+            if !set_name_before_spawn {
+                // Test scenario (1)
+                set_name_for_task(me(), TaskName::from("Child"));
+            }
+            shuttle::future::yield_now().await;
+        })
+    });
+
+    for h in handles {
+        h.await.unwrap();
+    }
+}
+
+#[test]
+fn test_tracing_with_label_fn() {
+    let metrics = RunnableSubscriber {};
+    let _guard = tracing::subscriber::set_default(metrics);
+
+    check_random(
+        || {
+            future::block_on(async { label_fn_inner(true).await });
+        },
+        10,
+    );
+}
+
+#[test]
+#[should_panic(expected = "assertion failed")]
+fn test_tracing_without_label_fn() {
+    let metrics = RunnableSubscriber {};
+    let _guard = tracing::subscriber::set_default(metrics);
+
+    check_random(
+        || {
+            future::block_on(async { label_fn_inner(false).await });
+        },
+        1, // even one execution is enough to fail the assertion
+    );
+}
+
+// Custom Subscriber implementation to monitor and check debug output generated by Shuttle.
+struct RunnableSubscriber;
+
+impl Subscriber for RunnableSubscriber {
+    fn enabled(&self, _metadata: &Metadata<'_>) -> bool {
+        true
+    }
+
+    fn new_span(&self, _span: &Attributes<'_>) -> Id {
+        // We don't care about span equality so just use the same identity for everything
+        Id::from_u64(1)
+    }
+
+    fn record(&self, _span: &Id, _values: &Record<'_>) {}
+    fn record_follows_from(&self, _span: &Id, _follows: &Id) {}
+
+    fn event(&self, event: &Event<'_>) {
+        let metadata = event.metadata();
+        let target = metadata.target();
+        if target.contains("shuttle") && target.ends_with("::runtime::execution") {
```

Comment:
Why is this no longer `target() ==` and why is it split up (why is it not `target.ends_with("shuttle::runtime::execution")`?)

Reply:
It provides a bit more flexibility, and doesn't make the check much weaker.